### PR TITLE
Support predownload image in ads

### LINKED ISSUE #1048: [feature request] Support image pre-download for the advanced daemonset worklord
**What would you like to be added**:
Support image pre-download for the advanced daemonset worklord.

**Why is this needed**:
It is neccessary for user to accelerate  the progress of daemon applications upgrade by using the image pre-download feature.

### PR DESCRIPTION
**Pull request by ABNER-1**
### Ⅰ. Describe what this PR does
Support predownload image in ads

### Ⅱ. Does this pull request fix one issue?

fixes #1048 
### Ⅲ. Describe how to verify it
create ads and change images of it, then an imagepulljob will be created.
If every pods have beed  updated, the imagepulljob will be deleted.

### Ⅳ. Special notes for reviews
I am  not  sure whether it is neccessary to limit the daemonset size.
I am so  sorry that  I did not found how to make the predownload image order correspond to the order of daemonset  update.

### CODE REVIEW

File: pkg/controller/daemonset/daemonset_predownload_image.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package daemonset
+
+import (
+	"context"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (ABNER-1) [reply pkg/controller/daemonset/daemonset_predownload_image.go:21]: @sonatype-lift help

**Comment by FillZpp** [comment]
@ABNER-1 The daemonset UT failed, could you fix it?

**Comment by ABNER-1** [comment]
> @ABNER-1 The daemonset UT failed, could you fix it?

The unittest init struct is complex and different from cloneset. 
I am reading the implements.
I will fix it when I think I can get it.

**Comment by ABNER-1** [comment]
Looks like these two failed pull_request test due to statefulset controller.
However, I have no code changes about it. @FillZpp

File: pkg/controller/daemonset/daemonset_controller.go [hunk 2]
```diff
@@ -390,6 +403,36 @@ func (dsc *ReconcileDaemonSet) syncDaemonSet(request reconcile.Request) error {
...
+			// get ads pre-download annotation
+			minUpdatedReadyPodsCount := 0
+			if minUpdatedReadyPods, ok := ds.Annotations[appsv1alpha1.ImagePreDownloadMinUpdatedReadyPods]; ok {
+				minUpdatedReadyPodsIntStr := intstrutil.Parse(minUpdatedReadyPods)
+				minUpdatedReadyPodsCount, err = intstrutil.GetScaledValueFromIntOrPercent(&minUpdatedReadyPodsIntStr, int(ds.Status.DesiredNumberScheduled), true)
+				if err != nil {
+					klog.Errorf("Failed to GetScaledValueFromIntOrPercent of minUpdatedReadyPods for %s: %v", request, err)
+				}
+			}
+			updatedReadyReplicas := ds.Status.UpdatedNumberScheduled
```
Comment (furykerry) [inline pkg/controller/daemonset/daemonset_controller.go:418]: UpdatedNumberScheduled may container replicas which are not ready, maybe we should add an extra status field such as updatedReady

Comment (ABNER-1) [reply pkg/controller/daemonset/daemonset_controller.go:418]: Yes,  you are right.
I will enable imagepredownload without checking updateReadyReplicas.

File: pkg/controller/daemonset/daemonset_controller.go [hunk 3]
```diff
@@ -127,7 +134,11 @@ func Add(mgr manager.Manager) error {
 	if !utildiscovery.DiscoverGVK(controllerKind) {
 		return nil
 	}
-
+	if !utildiscovery.DiscoverGVK(appsv1alpha1.SchemeGroupVersion.WithKind("ImagePullJob")) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.KruiseDaemon) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.PreDownloadImageForInPlaceUpdate) {
```
Comment (furykerry) [inline pkg/controller/daemonset/daemonset_controller.go:139]: plz add a separate feature gate for PreDownloadImageForDaemonSet that is disabled by default, predownload for ads is still lack some feature e.g. updateReadyPods check before predownload

**Review by FillZpp — APPROVED** [review]

**Pull request merged**